ext/mbstring/php_unicode.c

```diff
@@ -229,195 +228,144 @@ static inline unsigned php_unicode_tofold_simple(unsigned code, enum mbfl_no_enc
 	return code;
 }
 
-static inline void php_unicode_tolower_full(unsigned code, enum mbfl_no_encoding enc,
-	mbfl_convert_filter* next_filter) {
-	code = php_unicode_tolower_raw(code, enc);
-	if (UNEXPECTED(code > 0xffffff)) {
-		unsigned len = code >> 24;
-		const unsigned *p = &_uccase_extra_table[code & 0xffffff];
-		while (len--) {
-			(next_filter->filter_function)(*++p, next_filter);
-		}
-	} else {
-		(next_filter->filter_function)(code, next_filter);
-	}
-}
-
-static inline void php_unicode_toupper_full(unsigned code, enum mbfl_no_encoding enc,
-	mbfl_convert_filter* next_filter) {
-	code = php_unicode_toupper_raw(code, enc);
-	if (UNEXPECTED(code > 0xffffff)) {
-		unsigned len = code >> 24;
-		const unsigned *p = &_uccase_extra_table[code & 0xffffff];
-		while (len--) {
-			(next_filter->filter_function)(*++p, next_filter);
-		}
-	} else {
-		(next_filter->filter_function)(code, next_filter);
-	}
-}
-
-static inline void php_unicode_totitle_full(unsigned code, enum mbfl_no_encoding enc,
-	mbfl_convert_filter* next_filter) {
-	code = php_unicode_totitle_raw(code, enc);
-	if (UNEXPECTED(code > 0xffffff)) {
-		unsigned len = code >> 24;
-		const unsigned *p = &_uccase_extra_table[code & 0xffffff];
-		while (len--) {
-			(next_filter->filter_function)(*++p, next_filter);
-		}
-	} else {
-		(next_filter->filter_function)(code, next_filter);
-	}
-}
-
-static inline void php_unicode_tofold_full(unsigned code, enum mbfl_no_encoding enc,
-	mbfl_convert_filter* next_filter) {
-	code = php_unicode_tofold_raw(code, enc);
-	if (UNEXPECTED(code > 0xffffff)) {
-		unsigned len = code >> 24;
-		const unsigned *p = &_uccase_extra_table[code & 0xffffff];
-		while (len--) {
-			(next_filter->filter_function)(*++p, next_filter);
-		}
-	} else {
-		(next_filter->filter_function)(code, next_filter);
+static uint32_t *emit_special_casing_sequence(uint32_t w, uint32_t *out)
+{
+	unsigned int len = w >> 24;
+	const unsigned int *p = &_uccase_extra_table[w & 0xFFFFFF];
+	while (len--) {
+		*out++ = *++p;
 	}
+	return out;
 }
 
-struct convert_case_data {
-	mbfl_convert_filter *next_filter;
-	enum mbfl_no_encoding no_encoding;
-	int case_mode;
-	int title_mode;
-};
-
-static int convert_case_filter(int c, void *void_data)
+MBSTRING_API zend_string *php_unicode_convert_case(int case_mode, const char *srcstr, size_t in_len, const mbfl_encoding *src_encoding, int illegal_mode, int illegal_substchar)
```

Comment:
For the case mode, would it make sense to add an enum which contains all of the different modes?

And are the `illegal_mode` and ``illegal_substchar`` really ints? And maybe not bool, and a char? (Again I now very little about character encodings so the current API might be sensible)

Reply:
Hmm. It seems that introducing an `enum` for `illegal_mode` may require reorganizing the header files for mbstring. Depending on which header file I define it in, there are a lot of issues with the order of header file inclusion.

Reply:
I am just converting `illegal_substchar` to `uint32_t`.

For `illegal_mode`, I would like to handle this in the future. It seems to open a "can of worms" which is not directly related to this PR.